dv/cosim/spike_cosim.cc

```diff
@@ -0,0 +1,516 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "spike_cosim.h"
+#include "config.h"
+#include "decode.h"
+#include "devices.h"
+#include "log_file.h"
+#include "processor.h"
+#include "simif.h"
+
+#include <cassert>
+#include <iostream>
+#include <sstream>
+
+SpikeCosim::SpikeCosim(uint32_t start_pc, uint32_t start_mtvec,
+                       const std::string &trace_log_path, bool secure_ibex,
+                       bool icache_en)
+    : pending_iside_error(false), nmi_mode(false) {
+  FILE *log_file = nullptr;
+  if (trace_log_path.length() != 0) {
+    log = std::make_unique<log_file_t>(trace_log_path.c_str());
+    log_file = log->get();
+  }
+
+  processor =
+      std::make_unique<processor_t>("RV32IMC", "MU", DEFAULT_VARCH, this, 0,
+                                    false, log_file, nullptr, secure_ibex,
+                                    icache_en);
+
+  // TODO: Uncomment when rebasing on up to date spike
+  processor->set_mmu_capability(IMPL_MMU_SBARE);
+  processor->get_state()->pc = start_pc;
+  processor->get_state()->mtvec->write(start_mtvec);
+
+  if (log) {
+    processor->set_debug(true);
+    processor->enable_log_commits();
+  }
+}
+
+// always return nullptr so all memory accesses go via mmio_load/mmio_store
+char *SpikeCosim::addr_to_mem(reg_t addr) { return nullptr; }
+
+bool SpikeCosim::mmio_load(reg_t addr, size_t len, uint8_t *bytes) {
+  bool bus_error = !bus.load(addr, len, bytes);
+
+  bool dut_error = false;
+
+  // Incoming access may be an iside or dside access. Calculate 32-bit aligned
+  // address and current 32-bit aligned PC to help determine which.
+  uint32_t aligned_addr = addr & 0xfffffffc;
+  uint32_t aligned_pc = processor->get_state()->pc & 0xfffffffc;
+
+  if (pending_iside_error && (aligned_addr == pending_iside_err_addr)) {
+    // Check if the incoming access is subject to an iside error, in which case
+    // assume it's an iside access and produce an error.
+    pending_iside_error = false;
+    dut_error = true;
+  } else if (aligned_pc != aligned_addr && (aligned_pc + 4) != aligned_addr) {
+    // Otherwise check if the aligned PC matches with the aligned address or an
+    // incremented aligned PC (to capture the unaligned 4-byte instruction
+    // case). Assume a successful iside access if either of these checks are
+    // true, otherwise assume a dside access and check against DUT dside
+    // accesses.
+    dut_error = check_mem_access(false, addr, len, bytes);
+  }
+
+  return !(bus_error | dut_error);
```

Comment:
Nit: I think `||` is probably more common for booleans in C++ (and similarly below).